Marlin/Cartridge.cpp

```diff
@@ -0,0 +1,237 @@
+/**
+ * Cartridge.cpp - Used to keep track of the status of the cartridges.
+ * Author: Dan Thompson (danthompson41@gmail.com)
+ * Date: 2/19/2016
+ * Used to keep track of all information about material cartridges, for use
+ * in decision making in other parts of the program
+ * Copyright (C) 2016 Voxel8
+ */
+
+#include "Marlin.h"
+#include "Cartridge.h"
+
+#include "./temperature.h"  // for disable_all_heaters()
+#include "./stepper.h"      // for quickStop()
+#include "./language.h"     // for MSG_T_CARTRIDGE_REMOVED
+
+//===========================================================================
+//=============================== Definitions ===============================
+//===========================================================================
+
+#define NUMBER_OF_CARTRIDGES               (2)
+#define CARTRIDGE_REMOVAL_HYSTERESIS_COUNT (10)
+
+//===========================================================================
+//============================ Private Variables ============================
+//===========================================================================
+
+static bool cartridgePresent[NUMBER_OF_CARTRIDGES] = {false, false};
```

Comment:
Can a cartridge be both present and removed? Or not removed and not present? If feels like these are mutually exclusive and can be represented with only one of these arrays.